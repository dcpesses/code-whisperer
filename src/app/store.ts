import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import counterReducer from '../features/redux-counter/counterSlice';
import todosReducer from '../features/todos/todosSlice';

const reducer = {
  counter: counterReducer,
  todos: todosReducer
};
export const store = configureStore({ reducer });

export function getStoreWithState(preloadedState?: RootState) {
  return configureStore({ reducer, preloadedState });
}

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
