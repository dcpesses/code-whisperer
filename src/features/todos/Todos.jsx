import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { addTodo, editTodo, toggleComplete, deleteTodo } from './todosSlice';

import './todos.css';

const Todos = () => {
  const [text, setText] = useState('');
  const todos = useSelector((state) => state.todos);
  const dispatch = useDispatch();

  const [editId, setEditId] = useState('');
  const [isEditing, setEditing] = useState(false);

  const createTodoListItem = ({id, completed, text}) => (
    <li className="todo-list-item" key={id}>
      <span className="item-checkbox">
        <input type="checkbox" checked={completed} onChange={() => handleToggleComplete(id)} title={completedTodoText(completed)} />
      </span>
      <span className="item-content" style={{textDecoration: completedTodoStyle(completed)}}>
        {text}
      </span>
      <span className="item-actions">
        <button className="item-actions-edit" onClick={()=>onEditToggle(id, text)} title="Edit">
          ✎
        </button>
        <button className="item-actions-delete" onClick={() => handleDeleteTodo(id)} title="Delete">
          ✖
        </button>
      </span>
    </li>
  );

  const handleInputChange = (e) => {
    setText(e.target.value);
  };

  const handleAddTodo = () => {
    if (text) {
      dispatch(addTodo(text));
      setText('');
    }
  };

  const onEditToggle = ( id, content ) => {
    setEditing(true);
    setEditId(id);
    setText(content);
  };

  const handleEditTodo = () => {
    if (text) {
      dispatch(editTodo({id: editId, text}));
      setText('');
      setEditing(false);
    }
  };
  const handleEditCancel = () => {
    setText('');
    setEditing(false);
  };

  const handleToggleComplete = (id) => {
    dispatch(toggleComplete(id));
  };

  const handleDeleteTodo = (id) => {
    dispatch(deleteTodo(id));
  };

  const completedTodoStyle = (completed) => completed ? 'line-through' : 'none';
  const completedTodoText = (completed) => completed ? 'Mark Incomplete' : 'Mark Complete';

  return (
    <div className="todos">
      {
        isEditing ?
          <div className="add-todo">
            <input type="text" value={text} onChange={handleInputChange} placeholder="Edit Todo" />
            <button onClick={handleEditTodo}>
              Edit
            </button>
            <button onClick={handleEditCancel}>
              Cancel
            </button>
          </div>
          :
          <div className="add-todo">
            <input type="text" value={text} onChange={handleInputChange} placeholder="Enter a Todo..." />
            <button onClick={handleAddTodo}>
              Add Todo
            </button>
          </div>
      }

      <ul className={isEditing ? 'todos-list editing' : 'todos-list'}>
        {todos.map(createTodoListItem)}
      </ul>
    </div>
  );
};

export default Todos;
