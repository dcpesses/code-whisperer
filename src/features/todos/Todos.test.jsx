/* eslint-env jest */
// import { vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';

import { getStoreWithState } from '../../app/store';
import { Provider } from 'react-redux';
import Todos from './Todos';

describe('Todos', () => {
  let store;
  beforeEach(() => {
    store = getStoreWithState();
  });
  test('Should render as expected', () => {
    const {container} = render(
      <Provider store={store}>
        <Todos />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  test('Should add and display new todo list item', () => {
    render(
      <Provider store={store}>
        <Todos />
      </Provider>
    );
    const forminput = screen.getByPlaceholderText('Enter a Todo...', {selector: 'input'});
    const button = screen.getByText('Add Todo');

    expect(forminput).toHaveTextContent('');
    fireEvent.change(forminput, {target: {value: 'Howdy Doody'}});
    fireEvent.click(button);

    const listitem = () => screen.getByText('Howdy Doody', {selector: 'span'});
    expect(listitem).toBeDefined();
  });

  test('Should edit and update a todo list item', () => {
    render(
      <Provider store={store}>
        <Todos />
      </Provider>
    );
    const editbutton = screen.getAllByTitle('Edit')[0];
    fireEvent.click(editbutton);

    const forminput = screen.getByPlaceholderText('Edit Todo', {selector: 'input'});
    expect(forminput.value).toBe(store.getState().todos[0].text);

    fireEvent.change(forminput, {target: {value: 'Howdy Doody'}});
    const updatebutton = screen.getByText('Edit', {selector: 'button'});
    fireEvent.click(updatebutton);

    const listitem = screen.queryByText('Howdy Doody', {selector: 'span'});
    expect(listitem).toBeDefined();
  });

  test('Should edit and cancel updating a todo list item', () => {
    render(
      <Provider store={store}>
        <Todos />
      </Provider>
    );
    const editbutton = screen.getAllByTitle('Edit')[0];
    fireEvent.click(editbutton);

    const forminput = screen.getByPlaceholderText('Edit Todo', {selector: 'input'});
    expect(forminput.value).toBe(store.getState().todos[0].text);

    fireEvent.change(forminput, {target: {value: 'Howdy Doody'}});
    const updatebutton = screen.getByText('Cancel', {selector: 'button'});
    fireEvent.click(updatebutton);

    const listitem = screen.queryByText('Howdy Doody', {selector: 'span'});
    expect(listitem).toBeNull();
  });

  test('Should delete a todo list item', () => {
    render(
      <Provider store={store}>
        <Todos />
      </Provider>
    );
    const todoItemText = store.getState().todos[0].text;
    expect(screen.getByText(todoItemText, {selector: 'span'})).toBeDefined();

    const deletebutton = screen.getAllByTitle('Delete')[0];
    fireEvent.click(deletebutton);

    expect(screen.queryByText(todoItemText, {selector: 'span'})).toBeNull();
  });

  test('Should mark a todo list item as completed', () => {
    render(
      <Provider store={store}>
        <Todos />
      </Provider>
    );
    const todoItemText = store.getState().todos[0].text;
    const listitemContent = screen.queryByText(todoItemText, {selector: 'span'});
    expect(listitemContent.style.textDecoration).toBe('none');

    const checkbox = screen.getByTitle('Mark Complete', {selector: 'input'});
    fireEvent.click(checkbox);

    expect(listitemContent.style.textDecoration).toBe('line-through');
  });
});

