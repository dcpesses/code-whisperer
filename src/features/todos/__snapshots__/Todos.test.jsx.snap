// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todos > Should render as expected 1`] = `
<div>
  <div
    class="todos"
  >
    <div
      class="add-todo"
    >
      <input
        placeholder="Enter a Todo..."
        type="text"
        value=""
      />
      <button>
        Add Todo
      </button>
    </div>
    <ul
      class="todos-list"
    >
      <li
        class="todo-list-item"
      >
        <span
          class="item-checkbox"
        >
          <input
            title="Mark Complete"
            type="checkbox"
          />
        </span>
        <span
          class="item-content"
          style="text-decoration: none;"
        >
          Hit the gym
        </span>
        <span
          class="item-actions"
        >
          <button
            class="item-actions-edit"
            title="Edit"
          >
            ✎
          </button>
          <button
            class="item-actions-delete"
            title="Delete"
          >
            ✖
          </button>
        </span>
      </li>
      <li
        class="todo-list-item"
      >
        <span
          class="item-checkbox"
        >
          <input
            title="Mark Incomplete"
            type="checkbox"
          />
        </span>
        <span
          class="item-content"
          style="text-decoration: line-through;"
        >
          Meet George
        </span>
        <span
          class="item-actions"
        >
          <button
            class="item-actions-edit"
            title="Edit"
          >
            ✎
          </button>
          <button
            class="item-actions-delete"
            title="Delete"
          >
            ✖
          </button>
        </span>
      </li>
    </ul>
  </div>
</div>
`;
